File: notification-service/src/main/java/com/freelencerhub/notification/service/NotificationService.java
package com.freelencerhub.notification.service;

import com.freelencerhub.notification.dto.NotificationDTO;
import com.freelencerhub.notification.entity.Notification;
import com.freelencerhub.notification.repository.NotificationRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@Slf4j
public class NotificationService {

    private final NotificationRepository notificationRepository;

    @Autowired
    public NotificationService(NotificationRepository notificationRepository) {
        this.notificationRepository = notificationRepository;
    }

    public List<NotificationDTO> getAllNotifications() {
        return notificationRepository.findAll().stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    public List<NotificationDTO> getNotificationsByUser(Long userId) {
        return notificationRepository.findByUserId(userId).stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    public List<NotificationDTO> getNotificationsBySender(Long senderId) {
        return notificationRepository.findBySenderId(senderId).stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    @Async
    public void processNotificationAsync(Notification notification) {
        log.info("Processing notification asynchronously from sender: {} for user: {}", notification.getSenderId(), notification.getUserId());
        notification.setRead(false);
        notificationRepository.save(notification);
        log.info("Notification saved successfully.");
    }

    public void markAsRead(Long id) {
        notificationRepository.findById(id).ifPresent(n -> {
            n.setRead(true);
            notificationRepository.save(n);
        });
    }

    private NotificationDTO mapToDTO(Notification n) {
        return new NotificationDTO(n.getId(), n.getUserId(), n.getSenderId(), n.getMessage(), n.isRead(), n.getCreatedAt());
    }
}

File: frontend/src/services/notificationService.js
import api from './api';

const notificationService = {
  getAllNotifications: async () => {
    try {
      const response = await api.get('/notifications');
      return response.data;
    } catch (error) {
      console.error(error);
      throw error;
    }
  },

  getNotificationsByUser: async (userId) => {
    try {
      const response = await api.get(`/notifications/user/${userId}`);
      return response.data;
    } catch (error) {
      console.error(error);
      throw error;
    }
  },

  getNotificationsBySender: async (senderId) => {
    try {
      const response = await api.get(`/notifications/sender/${senderId}`);
      return response.data;
    } catch (error) {
      console.error(error);
      throw error;
    }
  },

  sendNotification: async ({ userId, senderId, message }) => {
    try {
      const response = await api.post('/notifications', { userId, senderId, message, isRead: false });
      return response.data;
    } catch (error) {
      console.error(error);
      throw error;
    }
  },
  
  markAsRead: async (id) => {
    try {
      const response = await api.put(`/notifications/${id}/read`);
      return response.data;
    } catch (error) {
      console.error(error);
      throw error;
    }
  },
};

export default notificationService;
